File: python/add_recorders.py
# python/add_recorders.py — recorder injection logic
# Adds NumpyArray recorders to all nodes that lack one.
# Recorder assignment rules from TASKS.md Task 17.
# Does NOT write to disk — returns the modified model dict and list of additions.

from typing import Any


# Recorder type constants — names from PYWR_SCHEMA.md (use NumpyArrayNodeDeficitRecorder)
RECORDER_NODE = "NumpyArrayNodeRecorder"
RECORDER_DEFICIT = "NumpyArrayNodeDeficitRecorder"
RECORDER_STORAGE = "NumpyArrayStorageRecorder"
RECORDER_NORMALISED = "NumpyArrayNormalisedStorageRecorder"

# Node types that get NumpyArrayStorageRecorder
STORAGE_TYPES = {"storage", "annualvirtualstorage"}

# Node types that get NumpyArrayNodeRecorder
NODE_RECORDER_TYPES = {
    "input", "output", "link", "river", "rivergauge", "catchment",
    "piecewiselink", "riversplithwithgauge",
}

# Virtual/aggregation types that do not get recorders via this route
SKIP_TYPES = {"aggregatednode", "aggregatedstorage", "virtualstorage"}


def _type_key(t: str) -> str:
    return t.lower()


def _recorder_name_already_exists(
    recorder_type: str,
    node_name: str,
    recorders: dict[str, Any],
) -> bool:
    """Return True if a recorder of the given type already targets node_name."""
    for rec_cfg in recorders.values():
        if (
            isinstance(rec_cfg, dict)
            and rec_cfg.get("type") == recorder_type
            and rec_cfg.get("node") == node_name
        ):
            return True
    return False


def add_recorders(model: dict[str, Any]) -> tuple[dict[str, Any], list[dict[str, str]]]:
    """
    Add appropriate NumpyArray recorders to all nodes that lack one.

    Returns:
        (modified_model, added_list)
        where added_list is a list of {"recorder_type": "...", "node": "..."} dicts.
    """
    import copy
    model = copy.deepcopy(model)

    nodes: list[dict[str, Any]] = model.get("nodes", [])
    recorders: dict[str, Any] = model.get("recorders", {})
    if not isinstance(recorders, dict):
        recorders = {}
        model["recorders"] = recorders

    added: list[dict[str, str]] = []

    for node in nodes:
        name = node.get("name")
        node_type = node.get("type", "")
        if not name or not node_type:
            continue

        tk = _type_key(node_type)

        if tk in SKIP_TYPES:
            continue

        if tk in STORAGE_TYPES:
            # NumpyArrayStorageRecorder
            if not _recorder_name_already_exists(RECORDER_STORAGE, name, recorders):
                rec_name = f"{name}_recorder"
                recorders[rec_name] = {"type": RECORDER_STORAGE, "node": name}
                added.append({"recorder_type": RECORDER_STORAGE, "node": name})

            # AnnualVirtualStorage also gets NumpyArrayNormalisedStorageRecorder
            if tk == "annualvirtualstorage":
                if not _recorder_name_already_exists(RECORDER_NORMALISED, name, recorders):
                    rec_name = f"{name}_normalised_recorder"
                    recorders[rec_name] = {"type": RECORDER_NORMALISED, "node": name}
                    added.append({"recorder_type": RECORDER_NORMALISED, "node": name})

        elif tk in NODE_RECORDER_TYPES:
            # NumpyArrayNodeRecorder for all flow nodes
            if not _recorder_name_already_exists(RECORDER_NODE, name, recorders):
                rec_name = f"{name}_recorder"
                recorders[rec_name] = {"type": RECORDER_NODE, "node": name}
                added.append({"recorder_type": RECORDER_NODE, "node": name})

            # Output nodes ending in _DC also get NumpyArrayNodeDeficitRecorder
            if tk == "output" and name.endswith("_DC"):
                if not _recorder_name_already_exists(RECORDER_DEFICIT, name, recorders):
                    rec_name = f"{name}_deficit_recorder"
                    recorders[rec_name] = {"type": RECORDER_DEFICIT, "node": name}
                    added.append({"recorder_type": RECORDER_DEFICIT, "node": name})

    model["recorders"] = recorders
    return model, added

File: electron/add_recorders.js
// electron/add_recorders.js
// Port of python/add_recorders.py — recorder injection logic in Node.js.
// Adds NumpyArray recorders to all nodes that lack one.
// Does NOT write to disk — returns { model, added }.

'use strict';

const RECORDER_NODE       = 'NumpyArrayNodeRecorder';
const RECORDER_DEFICIT    = 'NumpyArrayNodeDeficitRecorder';
const RECORDER_STORAGE    = 'NumpyArrayStorageRecorder';
const RECORDER_NORMALISED = 'NumpyArrayNormalisedStorageRecorder';

const STORAGE_TYPES = new Set(['storage', 'annualvirtualstorage']);

const NODE_RECORDER_TYPES = new Set([
  'input', 'output', 'link', 'river', 'rivergauge', 'catchment',
  'piecewiselink', 'riversplithwithgauge',
]);

const SKIP_TYPES = new Set(['aggregatednode', 'aggregatedstorage', 'virtualstorage']);

function tk(t) { return (t || '').toLowerCase(); }

function recorderAlreadyExists(recorderType, nodeName, recorders) {
  for (const rec of Object.values(recorders)) {
    if (rec && typeof rec === 'object' && rec.type === recorderType && rec.node === nodeName) {
      return true;
    }
  }
  return false;
}

function addRecorders(model) {
  // Deep clone so we don't mutate the caller's object
  model = JSON.parse(JSON.stringify(model));

  const nodes = Array.isArray(model.nodes) ? model.nodes : [];
  if (!model.recorders || typeof model.recorders !== 'object' || Array.isArray(model.recorders)) {
    model.recorders = {};
  }
  const recorders = model.recorders;
  const added = [];

  for (const node of nodes) {
    const name = node.name;
    const nodeType = node.type;
    if (!name || !nodeType) continue;

    const key = tk(nodeType);
    if (SKIP_TYPES.has(key)) continue;

    if (STORAGE_TYPES.has(key)) {
      if (!recorderAlreadyExists(RECORDER_STORAGE, name, recorders)) {
        recorders[`${name}_recorder`] = { type: RECORDER_STORAGE, node: name };
        added.push({ recorder_type: RECORDER_STORAGE, node: name });
      }
      // AnnualVirtualStorage also gets a normalised recorder
      if (key === 'annualvirtualstorage' && !recorderAlreadyExists(RECORDER_NORMALISED, name, recorders)) {
        recorders[`${name}_normalised_recorder`] = { type: RECORDER_NORMALISED, node: name };
        added.push({ recorder_type: RECORDER_NORMALISED, node: name });
      }
    } else if (NODE_RECORDER_TYPES.has(key)) {
      if (!recorderAlreadyExists(RECORDER_NODE, name, recorders)) {
        recorders[`${name}_recorder`] = { type: RECORDER_NODE, node: name };
        added.push({ recorder_type: RECORDER_NODE, node: name });
      }
      // Output nodes ending in _DC also get a deficit recorder
      if (key === 'output' && name.endsWith('_DC') && !recorderAlreadyExists(RECORDER_DEFICIT, name, recorders)) {
        recorders[`${name}_deficit_recorder`] = { type: RECORDER_DEFICIT, node: name };
        added.push({ recorder_type: RECORDER_DEFICIT, node: name });
      }
    }
  }

  model.recorders = recorders;
  return { model, added };
}

module.exports = { addRecorders };
